### DISCUSSION TITLE
Setting multiple server-side validation errors and focusing on first error

### DISCUSSION THREAD

**Discussion by anshumanb** [react-hook-form/react-hook-form#8842] [Q&A]
I have a form with multiple fields, where each field has client-side as well as server-side validation logic. All server-side validation is done on form submission only. (I've noted that we can have async custom validation rules for fields, but I'd like to server-side validate all the fields in one go, on form submission)

On submit, when there are multiple server-side validation errors (one per field), I'm iterating through the errors and using `setError` to mark fields in error. However, the current RHF API doesn't allow automatically focusing on the first error field, using `setError`. I have to manually calculate the first error field using something like (reduced example):

```js
const submit = async (submittedValues) => {
	try {
		await submitToServer(submittedValues);
	} catch (e) {
		if (e.status === 400) {
			const errorFields = e.payload.errors;
			const errorFieldNames = Object.keys(errorFields);
			// React-hook-form seems to order the keys as they appear
			// (or maybe the order in which they're registered). We use
			// this to figure out which field to focus on as there's no
			// better alternative available at the moment
			const orderedErrorFieldNames = Object.keys(submittedValues).filter((key) =>
				errorFieldNames.includes(key),
			);
			Object.entries(errorFields).forEach(([key, message]) => {
				setError(
					key,
					{ type: 'serverSide', message },
					// Focus on the first of the fields in error
					orderedErrorFieldNames[0] === key ? { shouldFocus: true } : undefined,
				);
			});
		}
	}
};

...

<form onSubmit={handleSubmit(submit)} ...>
```

My questions are:
* Is there something that will allow me to automatically focus on the first error field, like what happens with client-side validation errors, when using `setError` multiple times?
* I've assumed that keys in `submittedValues`, in my example, will always be ordered as they appear (or registered?). Is this a valid assumption and one that's likely to hold true in the future too?
* Is it worth having something in RHF to set multiple errors in one go while handling focus too? I've noted that [previously](https://github.com/react-hook-form/react-hook-form/pull/467) we had a `setErrors` although it didn't seem to handle focusing on the first error. Also, it was subsequently removed as thought to be confusing.

**Comment by bluebill1049** [comment]
automatic is probably not possible at the moment unless your server error is against a valid/registered input. so you will have to use https://react-hook-form.com/api/useform/setfocus API
